src/cmd/cmd.zig

```diff
@@ -989,75 +1000,70 @@ fn getOrDownloadSnapshots(
     allocator: Allocator,
     logger: Logger,
     gossip_service: ?*GossipService,
-) !SnapshotFieldsAndPaths {
+    // accounts_db_config: config.AccountsDBConfig,
+    options: struct {
+        snapshot_dir: std.fs.Dir,
```

Comment:
Passing files around as path strings is largely something I want to eliminate from the codebase, and that is a goal intentionally furthered by this PR.

Making this parameter into a path string moves the responsibility of taking care of that IO into this function, obfuscating it from the higher level, which hides the fact that this file is repeatedly opened and closed in various different contexts, creating many disparate and redundant syscalls.

As well, passing around string paths & returning the files they represent leads to a brittle memory management strategy for the string paths - `deinitPaths` was something I came up with and explicitly labelled as a hack, because the memory management model it represented was fundamentally fragile and easy to get wrong - though just slightly less wrong than the bug it replaced, which was caused by storing the paths as strings and having the original struct take ownership of both them and the snapshot fields, despite the snapshot field data actually having its own distinct lifetime in practice.

Overall, I am pretty strongly opposed to passing around path strings, and am confident that passing around file handles down from the highest level as a conservative resource handle is the most maintainable and optimal path moving forward.
If this rationale still leaves any doubt, I'd be happy to huddle so we can dedicate a proper discussion to this.

Reply:
I think I understand your perspective. I'm fine keeping it how you have it. I'm leaving this unresolved for visibility, but feel free to resolve this conversation.

I get the idea of instantiating the Dir in one high-level location, and then passing it around, rather than passing around a string and instantiating the same Dir in multiple places. I can see two rationales for that approach: better performance and centralized error handling.

But to accomplish those goals, you do not need to instantiate it at the highest level possible. I would take it case by case, but I think I would prefer not to elevate it beyond the level needed to centralize the instantiation. And that seems to be how you have it set up now, so it looks good to me. In some cases, it might make more sense for a lower context to decide how to handle various scenarios that occur when you try to open a dir. But I don't think that's an issue here. I just wouldn't move it up higher out of loadSnapshot. I think it makes more sense for higher levels to deal with configuration as just data (like a string or path struct).

Reply:
Yeah, I see your point. The ideal is going to be in some happy medium - current status quo of passing around string paths and concatenating them to open magical relative directories is pretty bad, but I'll agree that hoisting all directories as high up as possible would be the other terrible end of the spectrum.
As long as we make conscious decisions going forward about what layer we cross this threshold, I am content.